### PR TITLE
Fix Issue 18743 - ConditionalExpression with AssignExpression should require parens

### PR DESCRIPTION
**Pull request by ntrel**
Depends on https://github.com/dlang/phobos/pull/6505.

### CODE REVIEW

File: src/dmd/parse.d [hunk 1]
```diff
@@ -8483,6 +8483,11 @@ final class Parser(AST) : Lexer
         auto e = parseCondExp();
         while (1)
         {
+            // require parens for e.g. a ? a : a = a;
```
Comment (WalterBright) [inline src/dmd/parse.d]: This example doesn't seem to be in the test cases.

File: src/dmd/parse.d [hunk 2]
```diff
@@ -8483,6 +8483,11 @@ final class Parser(AST) : Lexer
...
+            if (precedence[e.op] == PREC.cond && !e.parens && precedence[token.value] == PREC.assign)
```
Comment (WalterBright) [inline src/dmd/parse.d]: Replace `precedence[e.op] == PREC.cond` with `e.op == TOK.question`

**Comment by jacob-carlborg** [comment]
If this requires a change in Phobos it means it's a breaking change and requires going through the regular deprecation phase.

File: src/dmd/parse.d (hunk 2 above)
Comment (ntrel) [reply src/dmd/parse.d]: Done.

File: src/dmd/parse.d (hunk 1 above)
Comment (ntrel) [reply src/dmd/parse.d]: I've now updated this line to be equivalent to a test case.

**Comment by ntrel** [comment]
@jacob-carlborg Added changelog, deprecation date comment and spec pull: https://github.com/dlang/dlang.org/pull/2366. Also addressed Walter's comments.

File: src/dmd/parse.d [hunk 3]
```diff
@@ -8483,6 +8483,14 @@ final class Parser(AST) : Lexer
         auto e = parseCondExp();
         while (1)
         {
+            // require parens for e.g. `t ? a = 1 : b = 2`
+            // Deprecated in 2018-05.
+            // Change to error in 2019-05.
+            // @@@DEPRECATED_2019-05@@@.
+            if (e.op == TOK.question && !e.parens && precedence[token.value] == PREC.assign)
+                dmd.errors.deprecation(e.loc, "`%s` must be parenthesized when next to operator `%s`",
+                    e.toChars(), Token.toChars(token.value));
```
Comment (jacob-carlborg) [inline src/dmd/parse.d]: I suggest that the deprecation message contains a suggestion on how to add the parentheses to get the same behavior as before the deprecation was added. The changelog should reflect this as well.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

**Review by jacob-carlborg — APPROVED** [review]
LGTM, although I still think it needs approval.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
ping @andralex @WalterBright. This PR is ready to merge. Are you against this addition?

**Review by wilzbach — APPROVED** [review]
FWIW I'm in favor of this. Thanks a lot for making this happen!
@WalterBright or @andralex can we make this happen?

**Review by Geod24 — APPROVED** [review]
LGTM

File: src/dmd/parse.d (hunk 3 above)
Comment (JinShil) [reply src/dmd/parse.d]: > assuming the reader knows what parenthesized means

I had to refer to the changelog example to understand what it meant.

Comment (jacob-carlborg) [reply src/dmd/parse.d]: Would "surrounded by parentheses" be better?

Comment (JinShil) [reply src/dmd/parse.d]: @andralex and @WalterBright aren't showing any interest in this, so I'm ready to move forward without their approval, but could you please consider @jacob-carlborg's suggestion with regard to the error message?

Comment (wilzbach) [reply src/dmd/parse.d]: > Edit: Made the change myself.

Ah I think we pushed in the same moment. Anyhow, I force-pushed my variant as the tests need updating too. Sorry.

**Comment by jacob-carlborg** [comment]
We could still include the fix in the error message or as a supplement message.

**Pull request merged**